tests/ga/test_extension.py

```diff
@@ -2535,304 +2536,285 @@ def test_ext_handler_with_systemd_cgroup_enabled(self, *args):
 
 
 class TestExtensionUpdateOnFailure(ExtensionTestCase):
+    
+    def setUp(self):
+        AgentTestCase.setUp(self)
+        self.mock_sleep = patch("time.sleep", lambda *_: mock_sleep(0.0001))
+        self.mock_sleep.start()
+
+    def tearDown(self):
+        self.mock_sleep.stop()
+        AgentTestCase.tearDown(self)
 
     @staticmethod
-    def _get_ext_handler_instance(name, version, handler=None, continue_on_update_failure=False):
+    def _do_upgrade_scenario(firstVersionActor, secondVersionActor):
+        """
+        Given the provided ExtensionActor objects, installs the first and then attempts
+        to update to the second. StatusBlobs and command invocations for each actor can
+        be checked with {actor}.statusBlobs and {actor}.get_command({command_name})
+        respectively.
+
+        Note that this method assumes the firstVersionActor's install command should
+        succeed. Don't use this method if your test is attempting to emulate a fresh install
+        (i.e. not an upgrade) with a failing installCommand.
+        """
+        
+        with get_protocol_and_handler(firstVersionActor) as (protocol, exthandlers_handler):
 
-        handler_json = {
-            "installCommand": "sample.py -install",
-            "uninstallCommand": "sample.py -uninstall",
-            "updateCommand": "sample.py -update",
-            "enableCommand": "sample.py -enable",
-            "disableCommand": "sample.py -disable",
-            "rebootAfterInstall": False,
-            "reportHeartbeat": False,
-            "continueOnUpdateFailure": continue_on_update_failure
-        }
+            # Run the handler with only the first version to install it
+            exthandlers_handler.run()
+            
+            # Verify the extension was picked up correctly.
+            # NOTE: this won't work for tests that want install to fail.
+            firstVersionActor.get_command("installCommand").assert_called_once()
+            firstVersionActor.get_command("enableCommand").assert_called_once()
```

Comment:
I'd throw rather than assert. This function seems to be a helper for test setup rather than verification of behavior of entity being tested.

Reply:
I'm not sure I understand. What exactly would be thrown?

Reply:
Sorry, i was trying to say asserts should be used to check the behavior you are testing, e.g " first_ext.get_command("disableCommand").assert_called_once()" in test_enabled_ext_should_be_disabled_at_ver_update

_do_upgrade_scenario is more of a setup method; it is preparing the extensions for the test to act on and do its checking. A setup failure should raise an exception rather than assert. In this case probably "raise Exception("The install command was not invoked when installing the extension")" or something similar would do it.

Reply:
That makes more sense. Thanks. I'll make those changes